README.TXT
WinClock(tm) v1.2 Copyright 1993 by Patrick Breen - All rights reserved.


Archive Contents:

   README.TXT    -  this file
   CHANGES.TXT   -  revision history of WinClock
   WINCLOCK.EXE  -  WinClock executable
   CLOCKDLL.DLL  -  WinClock support DLL


Description:

WinClock is a utility that displays useful information in the caption
bar of the active window.  This approach is somewhat less obtrusive
than having the information in a window which is always on top or
that can be covered by other windows.


System Requirements:

WinClock is completely self-contained and is compatable with both
Windows 3.0 and Windows 3.1.  WinClock uses very little memory (less
than 30k) and uses very few resources.


Installation:

Since you are reading this, I assume that you have unzipped the WinClock
archive.  The rest of the installation is very easy.  Create a new program
manager item for WinClock.  Make sure CLOCKDLL.DLL is in the working
directory or put it in your Windows directory.  Double-click to start
WinClock.  I prefer to add the program item to the Startup group in
Windows 3.1.  This causes WinClock to launch every time I start Windows.


Instructions:

Once WinClock is running, information will appear in the caption bar of the
active window.  Initially, WinClock displays the date on the left of the bar
and the time on the right.  A short click on either the time or the date will
toggle the display of free memory.  Another short click will display the
percentage of free resources available.  One more click will display the
disk space available on drive C.  Click and hold the mouse button to display
a menu.  The menu commands are as follows:

Setup...
	The setup dialog is used to specify how the information is displayed.
	The large group of radio buttons is used to control where each of
	the items appears.  An item can either appear on the left or right
	of the caption bar or it can be toggled into view.  If more than one
	item is specified for the left or right, they are combined and
	displayed together.  If more than one item is set to toggled, the
	first short click displays the first toggled item, a click on this
	item displays the next item, etc.  In addition, an item can be turned
	off which prevents it from being displayed at any time.  (If all items
	are turned off or all are set to toggle, the time will continue to
	display so the menu can be accessed!)

	In addition this dialog is used to set the time format and date format.
	The clock can be 24-hour and optionally display seconds.  The date
	format can be chosen from the set of formats displayed.

About...
	Provides the current version number and tells you where
	you can reach me.

Exit
	Removes WinClock and quits the application.


Advanced Options:

I have added the following advanced options (translation - there is no UI
to set them!) to modify the behavior of WinClock.  All of these options are
controlled by adding (or changing) lines in the WinClock.INI file which
can be found in your Windows directory.

ShowWindow=

This determines whether the WinClock window appears if the clock cannot
find a home.  By default, the window does appear and this value is set
to 1.  To keep the window from appearing, change this value to a 0.

SystemColors=

This determines whether WinClock uses the system colors to draw the
text or a custom color that from this .INI file.  By default, WinClock
uses the system colors and this value is set to 1.  To specify a custom
text color, change this value to a 0 and specify the color by adding the
following lines to the .INI file:
TextColorR=(red component of text color)
TextColorG=(green component of text color)
TextColorB=(blue component of text color)

EraseAlways=

This determines whether WinClock invalidates the desktop window to
remove "ghost" clocks in Windows 3.0.  For machines that are running
Windows 3.0, this option will eliminate the redraw that occurs when
switching between applications.  By default, WinClock does invalidate
and this value is set to 1.  This option has no effect if you are using
Windows 3.1.

LeftOffset=
RightOffset=

These two values specify how much room WinClock should leave between
the string it draws and the standard system controls.  These can be
useful if you are using other utilities that want to put controls in
the title bar.  By default, both values are set to 0.


Shareware:

Feel free to distribute this program to others.  All I ask is that this
file remain with the distributed copies.  This product is shareware.  If
you like it, send $5.  If you want to licence a number of copies, the
following table gives the discount rates:

Number of users	Cost		Price per user
  25				$ 100	$4.00
  50				$ 150	$3.00
 100				$ 250	$2.50
 250				$ 500	$2.00
 500				$ 750	$1.50
1000				$1000	$1.00


Disclaimer:

This software is distributed "as is".  Although I know of no problems
with the software, I make no warranties.


Other:

If you have any problems with this program or ideas for features that
should be added, I can be reached at the following electronic addresses
(in addition to the mail address and telephone number below).

Internet:    pbreen@world.std.com
CompuServe:  70312,743

___________________________ Order Form _________________________________


Send To:

Patrick Breen
3920 Mystic Valley Parkway #1119
Medford, MA 02155

Telephone:   (617) 396-2673


(  )  I enclose $5 for encouragement of continuing development



		  Name: ___________________________ Date:_______________


	    Company: ________________________________________________


         Address: ________________________________________________


City, State, Zip: ________________________________________________


         Country: ________________________________________________


       Day Phone: ___________________  Eve:  _____________________


Electronic Mail address: _________________________________________


How did you hear about WinClock v1.2? ____________________________


Comments:


CHANGES.TXT
WinClock(tm) Copyright 1993 by Patrick Breen - All rights reserved.


This file lists the changes between revisions of WinClock.


v1.2 - February 20, 1993

	Hopefully removed bug that caused system colors not to work
	on certain configurations.  I now default to using the active
	colors rather than the inactive colors.  (I was only able to
	reproduce and fis this using Windows 3.0 - if this is still a
	problem, let me know what version of Windows you are using)

	Completely tested and compatible with Windows 3.0.  There
	were a few compatability issues introduced with v1.1 that are
	now fixed.

	Removed custom colors - the clock now always uses system colors
	for the background and the text.  Removing this option also
	eliminated a bug which caused a "COMMDLG.DLG could not be found"
	message.  (Custom text color is still supported by options in the
	.INI file.)

	Make sure that the clock does not cover the entire caption
	bar on small windows.  Items on left are given preference over
	items on right.

	When all windows are iconized and WinClock is left homeless,
	information is displayed in the WinClock caption bar.  Can be
	disabled by setting in .INI file.

	Moved all preferences out of win.ini and into WinClock.ini.

	Added new display items for displaying free memory, free
	resources and free disk space on a chosen drive.

	Added options for what gets displayed when and where it gets
	displayed.  Setup dialog is now a hideous mass of radio buttons
	that allow the setting of these options.  Any suggestions on a
	better UI that uses standard controls would be appreciated.

	Added better support for international time and date settings.
	I now use the system settings for time separator, AM/PM text,
	and date separator.  In addition, DD/MM/YY and YY/MM/DD formats
	have been added.  Setup dialog now uses current date for sample.

	Added option to disable invalidation of entire desktop on
	machines running Windows 3.0.

	Once again, I thank all those who gave me ideas and described
	compatability problems.


v1.1 - February 7, 1993

	Fixed drawing problem that resulted in "ghost" clocks.  This
	version uses the Windows 3.1 function RedrawWindow() to invalidate
	the caption of the previous window which had the clock.  If
	RedrawWindow() is not available (ie, Windows 3.0 is being used),
	the entire desktop window is invalidated.  This does solve the
	problem, but it causes a noticeable flicker.  I will continue to
	look for a better solution.

	Fixed drawing problem in very small windows - the clock would
	cover the system menu and the window border.  Now it will never
	extend outside of the caption area.

	Added date.  A short click on the clock will display the date in
	the caption bar for a few seconds before displaying the time again.
	There are a couple of date formats available in the Setup... dialog.

	Removed flashing colon when seconds are not displayed.

	Added ability to position the clock on either the left or the
	right of the caption bar.  This is controlled by an option found
	in the Setup... dialog.

	Added Setup... dialog so that many options can be set at once.

	Added Color... dialog.  This dialog allows you to specify the
	text and background color of the clock.  It now defaults to using
	the system colors.  The Set... buttons are enabled if you have
	the Windows COMMDLG.DLL (I didn't want to create a fancy color
	dialog).  Otherwise, type in the RGB value of the color you desire.

	Thanks to all the people who contributed ideas and bug reports -
	I will continue to listen and respond to all contributions.


v1.0	- January 30, 1993

	Initial Release

WINCLOCK.H
#ifndef __WINCLOCK_H_
#define __WINCLOCK_H_

#include <windows.h>
#include <string.h>
#include <time.h>
#ifdef USE_FONT
#include <commdlg.h>
#endif

#define	NULLHWND		((HWND) 0)
#define	NULLHDC		((HDC) 0)
#define	NULLHMENU		((HMENU) 0)
#define	NULLHFONT		((HFONT) 0)
#define	NULLHBRUSH	((HBRUSH) 0)
#define	NULLHINSTANCE	((HINSTANCE) 0)

#define	IDN_APP		"WinClock"
#define	IDN_INI		"WinClock.ini"
#define	IDN_WINDOW	"ClockWin"

#define	IDI_APP		100

#define	ID_SHOWMENU	100
#define	ID_SETUP		101
#define	ID_FONT		102
#define	ID_ABOUT		103
#define	ID_EXIT		104

#define	ID_MENUSETUP	201
#define	ID_MENUFONT	202
#define	ID_MENUABOUT	203
#define	ID_MENUEXIT	204

#define	BUF_SIZE		128

typedef enum {
	ePosLeft,
	ePosRight,
	ePosToggle,
	ePosOff,
	ePosCnt
} Position;

typedef enum {
	eTypeNone 	= 0x0000,
	eTypeFirst 	= 0x0001,

	eTypeTime 	= eTypeFirst,
	eTypeDate 	= 0x0002,
	eTypeMem		= 0x0004,
	eTypeRes 		= 0x0008,
	eTypeDisk		= 0x0010,

	eTypeLast		= 0x0020,
	eTypeCnt		= 5
} ObjType;

typedef enum {
	eDateMDY,
	eDateDMY,
	eDateYMD,
	eDateAbbrev,
	eDateLong,
	eDateCnt
} DateFormat;

typedef enum {
	eTimeSeconds = 0x0001,
	eTime24Hour = 0x0002
} TimeOptions;

typedef struct {

	int left;				// Left objects
	int right;			// Right objects
	int toggle;			// Toggled objects

	short toggleCnt;		// Number of objects in toggle array

// Other information
	DateFormat dateFormat;
	TimeOptions timeOptions;
	short diskID;

     char sTime[2];
	char sDate[2];
	char sAMPM[2][5];

// Color information
	BOOL bTransparent;		// If TRUE, blend with caption bar
	COLORREF textColor;

	BOOL bShowWindow;		// If TRUE, show WinClock window if no others
	BOOL bEraseAlways;		// If TRUE, refresh all windows in 3.0

     short leftOffset;		// Additional left offset
	short rightOffset;		// Additional right offset

// The following are filled in at runtime
	HMENU hMenu;			// Handle to menu

// Handles to functions that we dynamically link to -
// these are NULL if the function does not exist
	DWORD (PASCAL FAR *lpfnGetHeapSpaces)(HANDLE);
	BOOL (PASCAL FAR *lpfnRedrawWindow)(HWND, const RECT FAR*, HRGN, UINT);

#ifdef USE_FONT
// Font information
	short fontSize;		// Point size
	short fontWeight;		// Used for bold
	BOOL bItalic;
	char fontName[LF_FACESIZE];

// The following are filled in at runtime
	HFONT hFont;			// Handle to font

	BOOL (PASCAL FAR *lpfnChooseFont)(CHOOSEFONT FAR *);
#endif

} BarData, FAR *PBarData;


// Main window procedure

LRESULT FAR PASCAL WndProc(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam);

// Dialog procedures

BOOL CALLBACK About(HWND hwndDlg, UINT msg, WPARAM wParam, LPARAM lParam);
BOOL CALLBACK Setup(HWND hwndDlg, UINT msg, WPARAM wParam, LPARAM lParam);


// DLL Functions

BOOL FAR _export InstallFilter(HWND, PBarData);
void FAR _export RemoveFilter(void);

void FAR _export DrawBar(void);
void FAR _export DisplayMenu(HWND);
void FAR _export GetString(ObjType type, char FAR *pBuf, short len);

// Message filter

DWORD FAR PASCAL GetMsgFilter(int nCode, WORD wParam, LPMSG pMsg);

#endif


WINCLOCK.C
/*
	WinClock(tm) v1.2

	Copyright (c) 1993  Patrick Breen
	All rights reserved.

	How to reach me:

		Patrick Breen
		3920 Mystic Valley Parkway #1119
		Medford, MA 02155

		(617) 396-2673

		Internet:		pbreen@world.std.com
		CompuServe: 	70312,743
*/

#include "winclock.h"

// Global data for all bar information
BarData barData;

// Application instance
HINSTANCE hBarInstance;

// Forward declaration of static functions
void NEAR ReadProfile(PBarData pData);
void NEAR WriteProfile(PBarData pData);
void NEAR SetProfileInt(char FAR *pSection, short n);

int PASCAL WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,
				LPSTR lpCmd, int nCmdShow)
{
	WNDPROC lpfnWndProc;
	WNDCLASS WindowClass;
	HINSTANCE hUser;
	HINSTANCE hKernel;
	HWND hBarWnd;
	MSG msg;

#ifdef USE_FONT
	HINSTANCE hCommDlg;
	OFSTRUCT of;
#endif

	// We can only have one instance running
	if (hPrevInstance != NULLHINSTANCE)
		return (0);

	// Set the global instance
	hBarInstance = hInstance;

	// Clear the data structure
	memset(&barData, 0, sizeof(BarData));

	// Load all dynamic functions that we need
#ifdef USE_FONT
	if ((OpenFile("commdlg.dll", &of, OF_EXIST) != HFILE_ERROR) &&
	    ((hCommDlg = LoadLibrary("COMMDLG.DLL")) > 32)) {

		(FARPROC) (barData.lpfnChooseFont) = GetProcAddress(hCommDlg, "ChooseFont");
	}
#endif

	if ((hUser = LoadLibrary("USER.DLL")) > 32) {

		(FARPROC) (barData.lpfnRedrawWindow) = GetProcAddress(hUser, "RedrawWindow");
	}

	if ((hKernel = LoadLibrary("KERNEL.DLL")) > 32) {

		(FARPROC) (barData.lpfnGetHeapSpaces) = GetProcAddress(hKernel, "GetHeapSpaces");
	}

	// Create the window procedure instance
	lpfnWndProc = (WNDPROC) MakeProcInstance((FARPROC) WndProc, hInstance);

	// If we could create the instance
	if (lpfnWndProc != NULL) {

		// Define the bar window class
		memset((void *) &WindowClass, 0, sizeof(WindowClass));
		WindowClass.lpfnWndProc   = lpfnWndProc;
		WindowClass.hInstance     = hInstance;
		WindowClass.hbrBackground = (HBRUSH) (COLOR_WINDOW + 1);
		WindowClass.lpszClassName = IDN_WINDOW;

		// Register the window class
		if (RegisterClass(&WindowClass)) {

			// Create the bar window
			hBarWnd = CreateWindow(IDN_WINDOW, "",
							   WS_OVERLAPPED | WS_SYSMENU | WS_THICKFRAME,
							   0, 0,
							   400, -4,
							   NULL, NULL,
							   hInstance, NULL );

			// If we could create the bar window
			if (hBarWnd != NULLHWND) {

				// Create the popup menu -
				// The contents of the menu is based upon
				// the availability of commdlg.dll - if this
				// DLL is present, extra options are added
				if ((barData.hMenu = CreatePopupMenu()) != NULLHMENU) {

					// Add the menu items to the menu
					AppendMenu(barData.hMenu, MF_ENABLED, ID_MENUSETUP, "Setup...");

#ifdef USE_FONT
					// Changing fonts is only supported if
					// the commdlg.dll exists - this saves
					// me from creating a font dialog
					if (barData.lpfnChooseFont != NULL) {
						AppendMenu(barData.hMenu, MF_ENABLED, ID_MENUFONT, "Font...");
					}

					// These items always exist
					AppendMenu(barData.hMenu, MF_SEPARATOR, 0, 0);
#endif

					AppendMenu(barData.hMenu, MF_ENABLED, ID_MENUABOUT, "About...");
					AppendMenu(barData.hMenu, MF_SEPARATOR, 0, 0);
					AppendMenu(barData.hMenu, MF_ENABLED, ID_MENUEXIT, "Exit");

					// Initialize the bar data
					ReadProfile(&barData);

					// Install the message filter that allows
					// us to detect a mouse down within the bar
					if (InstallFilter(hBarWnd, &barData)) {

						// Start the message loop
						while (GetMessage(&msg, NULL, NULL, NULL)) {
							TranslateMessage(&msg);
							DispatchMessage(&msg);
						}

					// Failed to install filter, back out
					} else {
						DestroyWindow(hBarWnd);
					}

					// Destroy the menu
					DestroyMenu(barData.hMenu);

				// Failed to create the menu, back out
				} else {
					DestroyWindow(hBarWnd);
				}
			}

			// All done with the class
			UnregisterClass(IDN_WINDOW, hInstance);
		}

		// Free the procedure instance
		FreeProcInstance((FARPROC) lpfnWndProc);
	}

	// Free all library handles
	if (hUser > 32) FreeLibrary(hUser);
	if (hKernel > 32) FreeLibrary(hKernel);
#ifdef USE_FONT
	if (hCommDlg > 32) FreeLibrary(hCommDlg);
#endif

	// Return something
	return (msg.wParam);
}


#ifdef USE_FONT
static void NEAR CacheFont(void)
{
	LOGFONT lf;

	memset(&lf, 0, sizeof(LOGFONT));

	lf.lfHeight = barData.fontSize;
	lf.lfItalic = barData.bItalic;
	lf.lfWeight = barData.fontWeight;
	strcpy(lf.lfFaceName, barData.fontName);

	if (barData.hFont != NULLHFONT)
		DeleteObject(barData.hFont);

	barData.hFont = CreateFontIndirect(&lf);
	return;
}


static void NEAR SetFont(void)
{
	CHOOSEFONT cf;
	LOGFONT lf;

	memset(&lf, 0, sizeof(LOGFONT));

	lf.lfHeight = barData.fontSize;
	lf.lfItalic = barData.bItalic;
	lf.lfWeight = barData.fontWeight;
	strcpy(lf.lfFaceName, barData.fontName);

	memset(&cf, 0, sizeof(CHOOSEFONT));

	cf.lStructSize = sizeof(CHOOSEFONT);
	cf.lpLogFont = &lf;
	cf.Flags = CF_INITTOLOGFONTSTRUCT | CF_SCREENFONTS;
	cf.nFontType = SCREEN_FONTTYPE;

	if ((barData.lpfnChooseFont != NULL) &&
	    (*(barData.lpfnChooseFont))(&cf)) {

		barData.fontSize = lf.lfHeight;
		barData.bItalic = lf.lfItalic;
		barData.fontWeight = lf.lfWeight;
		strcpy(barData.fontName, lf.lfFaceName);

		WriteProfile(&barData);
		CacheFont();
	}

	return;
}
#endif


LRESULT FAR PASCAL WndProc(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam)
{
	DLGPROC dlgprc;

	switch (msg) {

		case WM_CREATE:
			// Create a timer with a resolution of 1 second -
			// this is the guts of the clock
			SetTimer(hWnd, 1, 1000, NULL);
			break;

		case WM_TIMER:
			// Draw the contents of the bar
			DrawBar();
			break;

		case WM_COMMAND:
			// If this is a command
			// (either from the menu or one
			//  that we sent to ourselves)
			if (HIWORD(lParam) == 0) {

				// Switch on menu ID
				switch (wParam) {

					case ID_MENUSETUP:
					case ID_MENUABOUT:
					case ID_MENUEXIT:
						// Post ourselves a command so
						// that we don't crash when we
						// destroy ourselves
						PostMessage(hWnd, WM_COMMAND, wParam - 100, 0);
						break;

					case ID_SHOWMENU:
						// Show the menu to the user
						DisplayMenu(hWnd);
						break;

					case ID_ABOUT:
						// Show the about box
						dlgprc = (DLGPROC) MakeProcInstance(About, hBarInstance);
						DialogBox(hBarInstance, "About", NULLHWND, dlgprc);
						FreeProcInstance((FARPROC) dlgprc);
						break;

					case ID_SETUP:
						// Show the setup dialog
						dlgprc = (DLGPROC) MakeProcInstance(Setup, hBarInstance);
						DialogBox(hBarInstance, "Setup", NULLHWND, dlgprc);
						FreeProcInstance((FARPROC) dlgprc);
						break;

#ifdef USE_FONT
					case ID_FONT:
						// Show the font dialog
						// (available only if commdlg.dll is present)
						SetFont();
						break;
#endif

					case ID_EXIT:
						// Remove the message filter
						// and destroy the window
						RemoveFilter();
						DestroyWindow(hWnd);
						break;
				}

				return 0;
			}
			break;

		case WM_DESTROY:
			// Cause the entire desktop to redraw to remove
			// visual residue of ourselves
			InvalidateRect(NULLHWND, NULL, FALSE);

			// Dispose of the timer we created
			KillTimer(hWnd, 1);

			// And we are out of here!
			PostQuitMessage(0);
			return 0;
	}

	// Call DefProc
	return DefWindowProc(hWnd, msg, wParam, lParam);
}


BOOL CALLBACK About(HWND hwndDlg, UINT msg, WPARAM wParam, LPARAM lParam)
{
	BOOL bResult = FALSE;

	switch (msg) {

		case WM_COMMAND:
			// Any command message causes this
			// dialog to go awaw 
			EndDialog(hwndDlg, wParam);
			bResult = TRUE;
			break;
	}

	return bResult;
}


static Position NEAR GetObjectPos(PBarData pData, ObjType obj)
{
	Position pos = ePosOff;

	if (pData->left & obj)
		pos = ePosLeft;
	else if (pData->right & obj)
		pos = ePosRight;
	else if (pData->toggle & obj)
		pos = ePosToggle;

	return pos;
}


static void NEAR SetObjectPos(PBarData pData, ObjType obj, Position pos)
{
	// Turn bit off in all positions
	pData->left &= ~obj;
	pData->right &= ~obj;
	pData->toggle &= ~obj;

	// Turn on bit in proper position
	if (pos == ePosLeft)
		pData->left |= obj;
	else if (pos == ePosRight)
		pData->right |= obj;
	else if (pos == ePosToggle)
		pData->toggle |= obj;

	return;
}


static short NEAR GetRadioIdx(HWND hwnd, short base, short cnt)
{
	short i;

	for (i = 0; i < cnt; i++) {
		if (IsDlgButtonChecked(hwnd, base + i))
			break;
	}

	return i;
}


BOOL CALLBACK Setup(HWND hwndDlg, UINT msg, WPARAM wParam, LPARAM lParam)
{
	BOOL bResult = FALSE;
	char buf[BUF_SIZE];
	short i;

	switch (msg) {

		case WM_INITDIALOG:
			// Initialize the object positions
			for (i = 0; i < eTypeCnt; i++) {
				CheckDlgButton(hwndDlg,
							110 + (i * ePosCnt) +
							GetObjectPos(&barData, 1 << i), 1);
			}

			// Initialize the drive letter control
			buf[0] = (64 + barData.diskID);
			buf[1] = 0;
			SetDlgItemText(hwndDlg, 100, buf);

			// Initialize time format
			CheckDlgButton(hwndDlg, 102,
						((barData.timeOptions & eTime24Hour) != 0));
			CheckDlgButton(hwndDlg, 103,
						((barData.timeOptions & eTimeSeconds) != 0));

			// Initialize date format and date strings
			CheckDlgButton(hwndDlg, 105 + barData.dateFormat, 1);

			for (i = 0; i < eDateCnt; i++) {
				barData.dateFormat = i;
				GetString(eTypeDate, buf, BUF_SIZE);
				SetDlgItemText(hwndDlg, 105 + i, buf);
			}

			barData.dateFormat = GetRadioIdx(hwndDlg, 105, eDateCnt);
			break;

		case WM_COMMAND:

			switch (wParam) {

				case IDOK:
					// Update object positions
					for (i = 0; i < eTypeCnt; i++) {

						SetObjectPos(&barData, (1 << i),
								   GetRadioIdx(hwndDlg, 110 + (i * ePosCnt), ePosCnt));
					}

					// No valid items so we force
					// the time to display on the right
					if ((barData.left == eTypeNone) &&
					    (barData.right == eTypeNone))
						barData.right = eTypeTime;

					// Update drive letter
					GetDlgItemText(hwndDlg, 100, buf, 5);
					if ((buf[0] > 64) && (buf[0] < 92))
						barData.diskID = buf[0] - 64;

					// Update time options
					barData.timeOptions = 0;
					if (IsDlgButtonChecked(hwndDlg, 102))
						barData.timeOptions += eTime24Hour;
					if (IsDlgButtonChecked(hwndDlg, 103))
						barData.timeOptions += eTimeSeconds;

					// Update date format
					barData.dateFormat = GetRadioIdx(hwndDlg, 105, eDateCnt);

					// Write the profile
					WriteProfile(&barData);

					// Fall through...

				case IDCANCEL:
					EndDialog(hwndDlg, wParam);
					bResult = TRUE;
					break;
			}
			break;
	}

	return bResult;
}


static void NEAR UpdateToggleCnt(PBarData pData)
{
	ObjType type = pData->toggle;

	for (pData->toggleCnt = 0; type; type >>= 1) {
		if (type & 0x1) pData->toggleCnt++;
	}

	return;
}

static short NEAR ReadProfileInt(char FAR *pType, short def)
{
	return GetPrivateProfileInt(IDN_APP, pType, def, IDN_INI);
}


static void NEAR ReadProfile(PBarData pData)
{
	// Initialize the data from the users profile
	pData->left = ReadProfileInt("Left", eTypeDate);
	pData->right = ReadProfileInt("Right", eTypeTime);
	pData->toggle = ReadProfileInt("Toggle", eTypeRes | eTypeMem | eTypeDisk);

	pData->timeOptions = ReadProfileInt("TimeFormat", 0);
	pData->dateFormat = ReadProfileInt("DateFormat", eDateMDY);
	pData->diskID = ReadProfileInt("DiskID", 3);

	pData->leftOffset = ReadProfileInt("LeftOffset", 0);
	pData->rightOffset = ReadProfileInt("RightOffset", 0);

	pData->bShowWindow = ReadProfileInt("ShowWindow", 1);
	pData->bEraseAlways = ReadProfileInt("EraseAlways", 1);
	pData->bTransparent = ReadProfileInt("SystemColors", 1);

	if (!pData->bTransparent) {

		pData->textColor = RGB(ReadProfileInt("TextColorR", 0x00),
						   ReadProfileInt("TextColorG", 0x00),
						   ReadProfileInt("TextColorB", 0x00));
	}

#ifdef USE_FONT
	pData->fontSize = ReadProfileInt("FontSize", -13);
	pData->fontWeight = ReadProfileInt("Bold", 700);
	pData->bItalic = ReadProfileInt("Italic", 0);
	pData->hFont = NULLHFONT;

	GetPrivateProfileString(IDN_APP, "FontName", "System", pData->fontName, LF_FACESIZE, IDN_INI);

	CacheFont();
#endif USE_FONT

	UpdateToggleCnt(pData);

	// Read international date settings
	GetProfileString("intl", "sDate", "/", pData->sDate, 2);
	GetProfileString("intl", "sTime", ":", pData->sTime, 2);
	GetProfileString("intl", "s1159", "AM", pData->sAMPM[0], 5);
	GetProfileString("intl", "s2359", "PM", pData->sAMPM[1], 5);

	return;
}


static void NEAR SetProfileInt(char FAR *pType, short n)
{
	char Buf[BUF_SIZE];

	wsprintf(Buf, "%d", n);
	WritePrivateProfileString(IDN_APP, pType, Buf, IDN_INI);

	return;
}

static void NEAR WriteProfile(PBarData pData)
{
	SetProfileInt("Left", pData->left);
	SetProfileInt("Right", pData->right);
	SetProfileInt("Toggle", pData->toggle);

	SetProfileInt("TimeFormat", pData->timeOptions);
	SetProfileInt("DateFormat", pData->dateFormat);
	SetProfileInt("DiskID", pData->diskID);

	SetProfileInt("ShowWindow", pData->bShowWindow);
	SetProfileInt("SystemColors", pData->bTransparent);
	SetProfileInt("EraseAlways", pData->bEraseAlways);

	SetProfileInt("LeftOffset", pData->leftOffset);
	SetProfileInt("RightOffset", pData->rightOffset);

#ifdef USE_FONT
	SetProfileInt("FontSize", pData->fontSize);
	SetProfileInt("Bold", pData->fontWeight);
	SetProfileInt("Italic", pData->bItalic);

	WritePrivateProfileString(IDN_APP, "FontName", pData->fontName, IDN_INI);
#endif

	UpdateToggleCnt(pData);

	return;
}


